base/steer/FairRootManager.cxx

```diff
@@ -67,6 +70,24 @@ using std::list;
 using std::pair;
 using std::set;
 
+#if ( ROOT_VERSION_CODE >= ROOT_VERSION(6,9,3) )
+namespace {
+  TMCMutex createMutex2 = TMCMUTEX_INITIALIZER;
+  TMCMutex deleteMutex2 = TMCMUTEX_INITIALIZER;
+  TMCMutex registerMutex2 = TMCMUTEX_INITIALIZER;
+  TMCMutex fillMutex2  = TMCMUTEX_INITIALIZER;
+  TMCMutex tmpFillMutex2  = TMCMUTEX_INITIALIZER;
+  TMCMutex writeMutex2 = TMCMUTEX_INITIALIZER;
+  TMCMutex closeMutex2 = TMCMUTEX_INITIALIZER;
+  TMCMutex getMutex2 =  TMCMUTEX_INITIALIZER;
+}
```

Comment:
1. Could you please add some comments which document which shared data these mutexes are intended for? From the names they seem to be tied to the functions, which does not make much sense to us at the moment.

2. Is there any particular reason against using `std::mutex` (and `std::unique_lock`/`std::lock_guard` for the locks)?

Reply:
There is added a comment on the top of  mutexes definition; I have also changed the mutexes names as the '2'  was added in a different use case and makes no sense here.
This code is needed in the context of Geant4 MT simulation; there is used the same technology and locking method as in Geant4; see more in:
http://geant4.web.cern.ch/geant4/UserDocumentation/UsersGuides/ForToolkitDeveloper/html/ch02s14.html
It is planned to move to C++11 thread functions in Geant4 next year, so my plan is to follow then this development in Geant4 VMC. So for the time being I prefer to keep the same technique.

Reply:
Can all these mutexes be replaced just by a single one? Or am I missing something?

Reply:
To illustrate a bit more: Access to the member `TTree* FairRootManager::fOutTree` is synchronized with different mutexes, for example `fillMutex`, `fillTmpMutex`, and `deleteMutex`. This does not make sense to us. This implementation will still have data races, if one thread call `FairRootManager::FillWithLock` and another `FairRootManager::FillTmpLock` at the same time.

Unless, of course, you make sure that this never happens from the caller's site. But if this external synchronisation is in place, then you can just use a single mutex in all functions without a penalty (and it would be thread-safe without external synchronisation).

The only reason to have multiple mutexes really is, if you have independent data members that can be locked separately.

Reply:
First, please, note that in MT applications, there is one Root file and tree defined per thread, so we do not access the same three object by the operation, but its Root internal globals which gets modified in these operations. The problem with a global lock is that we get easily dead-locks. This mechanism of very local locks works fine in tests.
And also, the functions FillWithLock() and FillWithTmpLock() are private, the only public functions are Fill() and FillImpl(). When using just FillImpl(), no locks are applied; this function is used in MC application when running with sequential MCs.

Reply:
Just for my curiosity, do you know which Root internal globals are not thread-safe? Is there any chance we could apply the locking further down the call stack to achieve even better lock locality?

What percentage of cycles used by `TTree::Fill()` are actually spent on accessing the Root internal globals? I would imagine we are quite wasteful by locking the whole function call at FairRoot level, aren't we?